TheThingsNetwork/lorawan-stack
# Optimize Packet Broker Agent uplink token serialization

#### Summary



This PR refactors the serialization procedure for PBA-generate uplink tokens. The uplink tokens historically have been JSON encoded and encrypted using JSON web encryption. The refactor proposed here wraps the messages using pre-defined Protobuf messages and manually encrypts them in a manner similar to the existing solution (AES128/256 GCM).

The legacy format is kept for deserialization purposes - this is relevant for class C scheduling via PB, which may reuse an 'old format' uplink token for downlink scheduling. The legacy method should be removed in a subsequent minor release.

Example new token (158 bytes):
```
Cn8Kb99Zsmw+y7nzw4XqpwOoa+qq47UVC/C1SYWxBVGrufJ2RQSl7vxMOfZT/swnIkA5yt56S+FDMYA9nhqM3Jkd/v4CzQUqjbTcMH8ZIabYy0bbSGS6Mjd9Yth2KCnIqxpeG540jV7C1qULu7Lmotu2MxIMeOUDURlfavcaLQrrGi8KAwAAExIDdHRpGiNzdGFnaW5nMS5jbG91ZC50aGV0aGluZ3MuaW5kdXN0cmllcw==
```

Example old token (592 bytes):
```
eyJnIjoiWlhsS2FHSkhZMmxQYVVwQ1RWUkpORkl3VGs1VE1XTnBURU5LYkdKdFRXbFBhVXBDVFZSSk5GSXdUazVKYVhkcFlWaFphVTlwU21wTlJXaFFZbXN4ZEdGNlNtaFhSRUV3V1ROb1NVbHBkMmxrUjBadVNXcHZhVTVYVG5kV2JFNVdVVEpzY0ZSVVVtWlhSemxDVG01YWVWbHNiR2hhZVVvNUxtcGxiMmw0YVY5c2VERXdRVnBYVmpCNmFtUnFiSGN1V1dsVlMzSmxhMVZTV0RGWFF6bE1RaTU1U214dlVtdzVkWEExTFRnNFdGUlZkM1p6YW5wRFZITlZWRzFGYTAxNmFtcHBabE16Y2pKRmMwOUxkVTQwWld4NlYwWmhlbW80T1VJNFdtTlpkbmxIV21wWlozQmZjRGxETm5OdVkySXRkWGhOZDFKclYwRndWMEpVTjNFdFltbzJjRkJZTjNnM01FcG9ja0p1VVRadGNpMU5OV1ZxYVY4dFExcFBZVzVYUjFWSWFUSndabWxmWkhFMWRsWnBRa1kzZDBaM2IyUlRNWFp4VWxKcmVIRkNTbnBVVTNaeFkwVmtPRU5NVGw5d1NFaHJYMnRUT1RVM1JEVTNjalp3VWxrdU1WRjFTR3BzWWpWQ1RsVm5UbWxGYlVGRlYyRjJkdz09IiwiYSI6eyJmbmlkIjoiMDAwMDEzIiwiZnRpZCI6InR0aSIsImZjaWQiOiJzdGFnaW5nMS5jbG91ZC50aGV0aGluZ3MuaW5kdXN0cmllcyJ9fQ==
```

#### Changes



- Add PBA uplink token messages.
- Serialize gateway uplink tokens using a Protobuf message.
- Encrypt the gateway uplink tokens using AES128/256-GCM, in a similar manner to the existing JSON web encryption. The ciphertext and nonce are stored using a Protobuf message.
- Wrap the compound (gateway + forwarder + agent) uplink token using a Protobuf message.

#### Testing



Unit testing and `staging1`.

Manual test 1:
1. Have a gateway in tenant 1, and an end device registered in tenant 2.
2. Have the end device in tenant 2 join via the gateway in tenant 1, and also send an uplink and receive an application downlink. It suffices to check that the downlink has been scheduling in the gateway logs.

Manual test 2 (to be done by us - done on `staging1`):
1. Have a gateway in tenant 1, and an end device registered in tenant 2, marked as a class C device (doesn't have to be supported by the end device).
2. Have the end device send an uplink before this patch is applied.
3. Attempt to schedule a class C downlink. It suffices to check that it was scheduled on the gateway.

##### Regressions



As the legacy de-serialization method is kept, no regressions should occur.

#### Checklist
